### ISSUE TITLE
Large twoskip mailbox.db long search operations - performance issues

### ISSUE THREAD

**Issue by kwiatek6324** [cyrusimap/cyrus-imapd#3762]
It's a copy of issue on cyrus-users mailing list with additional tests.

Affected and tested on all 3.x versions (tested 3.4.2 / 3.2.8/3.0.15/3.0.0 )

Testing platform: i8700/32GBRam/Nvme storage, Debian stable,
But problems exists also on Centos8.x

**Problem:** The time spent on searching in mailbox is depending on records count, and affects performance because of fnctl locking problems

Collected data for reconstruct -rf (as example). Count of fnctl locks is a count of operations 
`
fcntl(3, F_SETLKW, {l_type=F_RDLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = 0
fstat(3, {st_mode=S_IFREG|0600, st_size=19975928, ...}) = 0
stat("/opt/store/usermailboxes//config/mailboxes.db", {st_mode=S_IFREG|0600, st_size=19975928, ...}) = 0
fcntl(3, F_SETLKW, {l_type=F_UNLCK, l_whence=SEEK_SET, l_start=0, l_len=0}) = 0
`
while reconstructing sample empty mailbox.

- 1000 mailboxes, reconstruct time: 0,030, fcntl count: 30
- 5000 mailboxes, reconstruct time: 0,016, fcntl count: 160
- 10000 mailboxes, reconstruct time: 0,068, fcntl count: 477
- 25000 mailboxes, reconstruct time: 0,124, fcntl count: 1178
- 50000 mailboxes, reconstruct time: 0,230, fcntl count: 2348
- 100000 mailboxes, reconstruct time: 0,398, fcntl count: 4682
- 200000 mailboxes, reconstruct time: 0,721, fcntl count: 9351
- 400000 mailboxes, reconstruct time: 1,328, fcntl count: 18637
- 550000 mailboxes, reconstruct time: 1,967, fcntl count: 25694

With 550k mailboxes on production reconstruct lasts for almost 4 seconds (empty mailbox) and fcntl count was about 27k.

Tested also skiplist and sql but always reconstruct time was about 4s with 550k mailboxes.
With 2.5 version problem does not exists. On test enviroment with Centos 8
reconstruct lasts 0.206s with full 550k records in mailboxes.db

**Comment by dilyanpalauzov** [comment]
You can try `./configure --with-lock=flock`.  This changes `fcntl(2)` with `flock(2)`.

**Comment by kwiatek6324** [comment]
Unable to compile:

libtool: link: gcc -g -O2 -o imap/.libs/imapd imap/imap_proxy.o imap/imapd.o imap/mutex_fake.o imap/proxy.o imap/sync_support.o master/service.o  -licui18n -licuuc -licudata -L/opt/store/sasl/lib imap/.libs/libcyrus_imap.so sieve/.libs/libcyrus_sieve.so lib/.libs/libcyrus.so lib/.libs/libcyrus_min.so /opt/store/sasl/lib/libsasl2.so -lssl -lcrypto com_err/et/.libs/libcyrus_com_err.so -ljansson -lpcre -lpcreposix -lz -Wl,-rpath -Wl,/opt/store/imapd-3.4.2/lib -Wl,-rpath -Wl,/opt/store/sasl/lib
/usr/bin/ld: imap/.libs/libcyrus_imap.so: undefined reference to `debug_locks_longer_than'

**Comment by kwiatek6324** [comment]
@dilyanpalauzov That was clue, but why:

I was able to compile after adding: 

EXPORTED double debug_locks_longer_than = 0.0;

to lib/lock_flock.c.

Timing with 550k mailboxes was:

real	0m0,020s
user	0m0,014s
sys	0m0,006s


and number of lock operations 10.

But why there is a problem with fcntl ?

**Comment by dilyanpalauzov** [comment]
> and number of lock operations 10.

Do you mean, that the sum of fcntl() and flock() calls numbers is 10?

**Comment by kwiatek6324** [comment]
> > and number of lock operations 10.
> 
> Do you mean, that the sum of fcntl() and flock() calls numbers is 10?

Not exactly. On the same environment as above, but compiled with "--with-lock=flock"
and mailboxes.db with 550000 entries, there were 18 flock calls, and reconstruct time 0,02s.

The same but without flock gives reconstruct time: 1,967 and  fcntl count: 25694.

**Comment by kwiatek6324** [comment]
Sorry for errors because of problem with my test environment.
Cyrus compiled with flock gives the same: 34000 flock syscalls and reconstruct time 2,2s

So problem exists in both cases.
